source/uwp/Renderer/lib/AdaptiveImageRenderer.cpp

```diff
@@ -66,517 +46,372 @@ namespace AdaptiveCards::Rendering::Uwp
-                                        IAsyncAction* /*action*/, AsyncStatus status) -> HRESULT {
-                                        if (status == AsyncStatus::Completed)
+                                setSourceAction.Completed(
+                                    [weakThis, uiElement, isAutoSize, parentElement, imageContainer, isVisible](winrt::IAsyncAction const&,
+                                                                                                      winrt::AsyncStatus status)
+                                    {
+                                        if (status == winrt::AsyncStatus::Completed && isAutoSize)
                                         {
-                                            // Here should be the auto resizing, at this time we already have the image and everything set
-                                            if (isAutoSize)
+                                            if (auto strongThis = weakThis.get())
                                             {
-                                                SetAutoSize(strongImageControl.Get(), parentElement, imageContainer, isVisible, false /* imageFiresOpenEvent */);
+                                                strongThis->SetAutoSize(uiElement, parentElement, imageContainer, isVisible, false /* imageFiresOpenEvent */);
                                             }
                                         }
-
-                                        return S_OK;
                                     });
-
-                                RETURN_IF_FAILED(setSourceAction->put_Completed(callbackSource.Get()));
-
-                                return S_OK;
                             }
-                            else
+                        }
+                        else
+                        {
+                            if (auto strongThis = weakThis.get())
                             {
-                                m_imageLoadTracker.MarkFailedLoadBitmapImage(bitmapImage.Get());
-                                return S_OK;
+                                this->m_imageLoadTracker->MarkFailedLoadBitmapImage(bitmapImage);
                             }
-                        })
-                        .Get()));
-
-                return;
+                        }
+                    });
             }
         }
 
-        INT32 isDataUriImage{};
-        THROW_IF_FAILED(WindowsCompareStringOrdinal(schemeName.Get(), HStringReference(L"data").Get(), &isDataUriImage));
-        if (isDataUriImage == 0)
+        if (schemeName == L"data")
         {
             // Decode base 64 string
-            HString dataPath;
-            THROW_IF_FAILED(imageUrl->get_Path(dataPath.GetAddressOf()));
-
-            std::string data = AdaptiveBase64Util::ExtractDataFromUri(HStringToUTF8(dataPath.Get()));
+            winrt::hstring dataPath = imageUrl.Path();
+            // TODO: look at this routine (not too sure about memcpy_s)
```

Comment:
@paulcam206 could you look at this please

Reply:
maybe this is a good use case for [`DataWriter`](https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/std-cpp-data-types)[(ref)](https://docs.microsoft.com/en-us/uwp/api/windows.storage.streams.datawriter?view=winrt-22000)?